_tour/variances.md

```diff
@@ -32,110 +36,90 @@ case class Cat(name: String) extends Animal
+printAnimalNames(cats)
 
-### Other Examples
+// prints: Fido, Rex
+printAnimalNames(dogs)
+```
 
-Another example that can help one understand variance is `trait Function1[-T, +R]` from the Scala standard library. `Function1` represents a function with one parameter, where the first type parameter `T` represents the parameter type, and the second type parameter `R` represents the return type. A `Function1` is contravariant over its parameter type, and covariant over its return type. For this example we'll use the literal notation `A => B` to represent a `Function1[A, B]`.
+### Contravariance
 
-Assume the similar `Cat`, `Dog`, `Animal` inheritance tree used earlier, plus the following:
+We've seen we can accomplish Covariance by making sure that we can't put something in the Covariant type, but only get something out. What if we had the opposite, something you can put something in, but can't take out? This situation arises if we have something like a serializer, that takes values of type A, and converts them to a serialized format.
 
 ```scala mdoc
-abstract class SmallAnimal extends Animal
-case class Mouse(name: String) extends SmallAnimal
+  abstract class Serializer[-A] {
+    def serialize(a: A): String
+  }
+
+  val animalSerializer: Serializer[Animal] = new Serializer[Animal] {
+    def serialize(animal: Animal): String = s"""{ "name": "${animal.name}" }""" 
+  }
+  val catSerializer: Serializer[Cat] = animalSerializer
+  catSerializer.serialize(Cat("Felix"))
 ```
 
-Suppose we're working with functions that accept types of animals, and return the types of food they eat. If we would like a `Cat => SmallAnimal` (because cats eat small animals), but are given a `Animal => Mouse` instead, our program will still work. Intuitively an `Animal => Mouse` will still accept a `Cat` as an argument, because a `Cat` is an `Animal`, and it returns a `Mouse`, which is also a `SmallAnimal`. Since we can safely and invisibly substitute the former with the latter, we can say `Animal => Mouse` is a subtype of `Cat => SmallAnimal`.
+We say that `Serializer` is *contravariant* in `A`, and this is indicated by the `-` before the `A`. A more general serializer is a subtype of a more specific serializer.
+
+More formally, that gives us the reverse relationship: given some `class Contra[-T]`, then if `A` is a subtype of `B`, `Cov[B]` is a subtype of `Cov[A]`. 
 
 ### Comparison With Other Languages
 
 Variance is supported in different ways by some languages that are similar to Scala. For example, variance annotations in Scala closely resemble those in C#, where the annotations are added when a class abstraction is defined (declaration-site variance). In Java, however, variance annotations are given by clients when a class abstraction is used (use-site variance).
+
+Scalas tendency towards immutable types makes it that covariant and contravariant types are more common than in other languages, since a mutabable generic type can't be covariant or contravariant.
```

Comment:
s/Scalas/Scala's/
s/mutabable/mutable/
s/can't be covariant or contravariant/must be invariant/